observer/nodesProviderFactory.go

```diff
@@ -0,0 +1,58 @@
+package observer
+
+import (
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-proxy-go/config"
+)
+
+var log = logger.GetOrCreate("observer")
+
+// nodesProviderFactory handles the creation of an nodes provider based on config
+type nodesProviderFactory struct {
+	cfg config.Config
+}
+
+// NewNodesProviderFactory returns a new instance of nodesProviderFactory
+func NewNodesProviderFactory(cfg config.Config) (*nodesProviderFactory, error) {
+	return &nodesProviderFactory{
+		cfg: cfg,
+	}, nil
+}
+
+// CreateObservers will create and return an object of type NodesProviderHandler based on a flag
+func (npf *nodesProviderFactory) CreateObservers() (NodesProviderHandler, error) {
+	if npf.cfg.GeneralSettings.BalancedObservers {
+		return NewCircularQueueNodesProvider(npf.cfg.Observers)
+	}
+
+	return NewSimpleNodesProvider(npf.cfg.Observers)
+}
+
+// CreateObservers will create and return an object of type NodesProviderHandler based on a flag
+func (npf *nodesProviderFactory) CreateFullHistoryNodes() (NodesProviderHandler, error) {
+	if npf.cfg.GeneralSettings.BalancedFullHistoryNodes {
+		nodesProviderHandler, err := NewCircularQueueNodesProvider(npf.cfg.FullHistoryNodes)
+		if err != nil {
+			return returnDisabledFullHistoryNodesProviderIfNeeded(err)
+		}
+
+		return nodesProviderHandler, nil
+	}
+
+	nodesProviderHandler, err := NewSimpleNodesProvider(npf.cfg.FullHistoryNodes)
+	if err != nil {
+		return returnDisabledFullHistoryNodesProviderIfNeeded(err)
+	}
+
+	return nodesProviderHandler, nil
+}
+
+func returnDisabledFullHistoryNodesProviderIfNeeded(err error) (NodesProviderHandler, error) {
```

Comment:
`get` instead of `return` perhaps?

Anyway, you can do this logic inline, in function `CreateFullHistoryNodes` (optional).

Reply:
renamed. left as it is because it is used in 2 places